### PR TITLE
⚡️ Improve state management with Twake router #1734 

### PR DESCRIPTION
**Pull request by tissarni**
Creating three hooks (`useRouterCompany`, `useRouterWorkspace` and `useRouterChannel`) in order to fragment `useRouteState` already existing and to used in Twake.

## Description

Those three hooks will return the id of the instance asked (ie : `companyId`) from the route.

## Related Issue

Improve state management with Twake router #1734 : https://app.zenhub.com/workspaces/twake-5f06c5bf7796310019354875/issues/linagora/twake/1734

## Motivation and Context

For now Twake render to many components that don't need to be re-render.

## Videos showing the improvement 

https://user-images.githubusercontent.com/46565431/142031505-551c382e-bb22-487b-a44d-c7bdc14b1d00.mov

https://user-images.githubusercontent.com/46565431/142031516-4277a4a6-fd70-42a4-825c-54f5c07a8095.mov

### CODE REVIEW

File: twake/frontend/src/app/scenes/Client/MainView/MainView.tsx [hunk 1]
```diff
@@ -9,18 +9,23 @@ import ChannelsBarService from 'app/services/channels/ChannelsBarService';
 import { useWatcher } from 'app/services/Observable/Observable';
 import AccountStatusComponent from 'app/components/OnBoarding/AccountStatusComponent';
 import CompanyBillingBanner from 'app/components/OnBoarding/CompanyBillingBanner';
-import useRouteState from 'app/services/hooks/useRouteState';
 
 import './MainView.scss';
+import useRouterCompany from 'app/services/hooks/useRouterCompany';
```
Comment (stephanevieira75) [inline twake/frontend/src/app/scenes/Client/MainView/MainView.tsx:14]: We should follow the same import pattern

``` js
import React from "react";

// external imports
import Something from "somethingmodule";
...

// local imports
import SomethingElse from "app/components/SomethingElse";
...

// style imports
import "./style.scss";
```

File: twake/frontend/src/app/scenes/Client/Popup/WorkspaceParameter/Pages/WorkspacePartnerTabs/Members.tsx [hunk 2]
```diff
@@ -9,14 +9,15 @@ import UserService from 'services/user/UserService';
...
 import InitService from 'app/services/InitService';
-import RouterServices from 'services/RouterService';
 import { ChevronUp, ChevronsUp } from 'react-feather';
 import useBreakpoint from 'antd/lib/grid/hooks/useBreakpoint';
 import Api from 'app/services/Api';
 import ConsoleService from 'app/services/Console/ConsoleService';
 import WorkspaceService from 'services/workspaces/workspaces.js';
 import _ from 'lodash';
 import { delayRequest } from 'app/services/utils/managedSearchRequest';
+import useRouterCompany from 'app/services/hooks/useRouterCompany';
```
Comment (stephanevieira75) [inline twake/frontend/src/app/scenes/Client/Popup/WorkspaceParameter/Pages/WorkspacePartnerTabs/Members.tsx:19]: Same here please

File: twake/frontend/src/app/scenes/Client/WorkspacesBar/WorkspacesBar.tsx [hunk 3]
```diff
@@ -6,20 +6,24 @@ import { Layout } from 'antd';
 import { WorkspaceType } from 'app/models/Workspace';
 import Group from './Components/Group/Group';
 import Workspace from './Components/Workspace/Workspace';
-import useRouteState from 'app/services/hooks/useRouteState';
 import { useCompanyWorkspaces } from 'app/state/recoil/hooks/useCompanyWorkspaces';
 
 import './WorkspacesBar.scss';
+import useRouterCompany from 'app/services/hooks/useRouterCompany';
```
Comment (stephanevieira75) [inline twake/frontend/src/app/scenes/Client/WorkspacesBar/WorkspacesBar.tsx:12]: and here

File: twake/frontend/src/app/services/hooks/useRouterChannel.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { useState, useCallback, useEffect } from 'react';
+import RouterService from '../RouterService';
+
+export default function useRouterChannel() {
+  const channelIdFromRoute = RouterService.getStateFromRoute().channelId;
+
+  const [channelId, setchannelId] = useState<string | undefined>(channelIdFromRoute);
+  const handleStateFromRoute = useCallback(
+    function () {
```
Comment (stephanevieira75) [inline twake/frontend/src/app/services/hooks/useRouterChannel.ts:9]: You can use arrow function instead

File: twake/frontend/src/app/services/hooks/useRouterCompany.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { useState, useCallback, useEffect } from 'react';
+import RouterService from '../RouterService';
+
+export default function useRouterCompany() {
+  const companyIdFromRoute = RouterService.getStateFromRoute().companyId;
+
+  const [companyId, setCompanyId] = useState<string | undefined>(companyIdFromRoute);
+  const handleStateFromRoute = useCallback(
+    function () {
```
Comment (stephanevieira75) [inline twake/frontend/src/app/services/hooks/useRouterCompany.ts:9]: use arrow function

File: twake/frontend/src/app/services/hooks/useRouterWorkspace.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { useState, useCallback, useEffect } from 'react';
+import RouterService from '../RouterService';
+
+export default function useRouterworkspace() {
+  const workspaceIdFromRoute = RouterService.getStateFromRoute().workspaceId;
+
+  const [workspaceId, setworkspaceId] = useState<string | undefined>(workspaceIdFromRoute);
+  const handleStateFromRoute = useCallback(
+    function () {
```
Comment (stephanevieira75) [inline twake/frontend/src/app/services/hooks/useRouterWorkspace.ts:9]: here use arrow function too

**Review by stephanevieira75 — CHANGES_REQUESTED** [review]
Good job and welcome in the fabulous world of React 😄

**Comment by RomaricMourgues** [comment]
@tissarni on Monday please:
- resolves the conversations
- push your current code
You must commit more frequently (on a daily basis) 😉

File: twake/frontend/src/app/scenes/Client/Client.tsx [hunk 7]
```diff
@@ -23,23 +23,23 @@ import LoginService from 'app/services/login/LoginService';
...
+  const workspaceId = useRouterWorkspace(); 
   const [menuIsOpen, setMenuIsOpen] = useState(false);
   const user = useCurrentUser();
 
-  popupService.useListener(useState);
+  /*popupService.useListener(useState);
   Workspaces.useListener(useState);
   Languages.useListener(useState);
-  LoginService.useListener(useState);
+  LoginService.useListener(useState);*/
```
Comment (RomaricMourgues) [inline twake/frontend/src/app/scenes/Client/Client.tsx:42]: You need to put back this (all except workspaces service):
```
popupService.useListener(useState);
Languages.useListener(useState);
LoginService.useListener(useState);
```

File: twake/frontend/src/app/scenes/Client/ChannelsBar/ChannelsBar.tsx [hunk 8]
```diff
@@ -44,9 +48,12 @@ const LoadingChannels = () => {
 };
 
 export default () => {
-  const { companyId, workspaceId } = RouterServices.getStateFromRoute();
+  //const { companyId, workspaceId } = RouterServices.getStateFromRoute();
```
Comment (RomaricMourgues) [inline twake/frontend/src/app/scenes/Client/ChannelsBar/ChannelsBar.tsx:51]: Don't forget to remove comments that should be removed

**Review by RomaricMourgues — CHANGES_REQUESTED** [review]

File: twake/frontend/src/app/state/recoil/atoms/Router.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
+import { atom, selector, selectorFamily } from 'recoil';
+
+import { ClientStateType } from 'app/services/RouterService';
+
+export const RouterState = atom<ClientStateType | undefined>({
+  key: 'RouterState',
+  default: undefined,
+});
+
+// to be moved in ./selector
```
Comment (stephanevieira75) [inline twake/frontend/src/app/state/recoil/atoms/Router.ts:10]: 👍🏻

**Review by stephanevieira75 — APPROVED** [review]

File: twake/frontend/src/app/state/recoil/atoms/Router.ts (hunk 9 above)
Comment (RomaricMourgues) [reply twake/frontend/src/app/state/recoil/atoms/Router.ts:10]: @tissarni could you implement this comment before I can merge ?

**Pull request merged**